### DISCUSSION TITLE
Is it possible to use the `creator` parameter with `create_async_engine`?

### DISCUSSION THREAD

**Discussion by jrouly** [sqlalchemy/sqlalchemy#7783] [Usage Questions] [answered]
I'm running into a lot of weirdness trying to specify a DB connection lambda (via the `creator` argument) passed to the `create_async_engine` method.

I've got similar code working fine with the sync `create_engine` method:
```python
return create_engine(
    url="postgresql+psycopg2://",
    creator=lambda: PostgresResource._create_connection(pg_config),
)
```

where `_create_connection()` will refresh an access token and then invoke `psycopg2.connect()`.

However, the parallel code for async does not work:
```python
async def creator():
    connection = AsyncPostgresResource._create_connection(pg_config)
    return await connection

engine = create_async_engine(
    url="postgresql+asyncpg://",
    creator=creator,
)

sessionmaker(
    class_=AsyncSession, # I've tried without this parameter as well, and it seems to default to a sync Session.
    bind=self.engine
)
```

where the `_creator` method is (I've tried both):
* an async method
* a synchronous method (using `asyncio.run()`)

that refreshes the access token and invokes `asyncpg.connect()`.

Errors I've seen include:
```
# lots of these
'await' outside async function
```

```
conn = <coroutine object AsyncPostgresResource._create_engine.<locals>.creator at 0x1103a0e40>

    def connect(conn):
>       conn.await_(self.setup_asyncpg_json_codec(conn))
E       AttributeError: 'coroutine' object has no attribute 'await_'
```

and when trying to use `asyncio.run()` in `creator()` to create a blocking/sync `creator` function in `create_async_engine`:
```
sqlalchemy.ext.asyncio.exc.AsyncContextNotStarted: AsyncConnection context has not been started and object has not been awaited.
```

---

I've also tried using the newer `@events` API:
```python
engine = create_async_engine(
    url="postgresql+asyncpg://",
    connect_args=connect_args,
    # creator=sync_creator,
)

@event.listens_for(engine, "do_connect")
def provide_token(dialect, conn_rec, cargs, cparams):
    cparams["password"] = get_password()
```

and I instead get this error:
```
self = <sqlalchemy.event.base.AsyncConnectionEventsDispatch object at 0x10b3a7660>
name = 'do_connect'

    def __getattr__(self, name):
        # Assign EmptyListeners as attributes on demand
        # to reduce startup time for new dispatch objects.
        try:
            ls = self._empty_listeners[name]
        except KeyError:
>           raise AttributeError(name)
E           AttributeError: do_connect
```

**Comment by jrouly** [comment, marked as answer]
I got it working by replacing `creator=` with `connect_args` and then using the Events API. I was able to get the Events API working by using the synchronous engine.


This fails:
```python
@event.listens_for(engine, "do_connect")
def provide_token(dialect, conn_rec, cargs, cparams):
    cparams["password"] = get_password()
```

This works:
```python
@event.listens_for(engine.sync_engine, "do_connect")
def provide_token(dialect, conn_rec, cargs, cparams):
    cparams["password"] = get_password()
```

**Reply by zzzeek** [reply]
yes event hooks for now need to refer to the "Engine", this is described at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension although we are currently lacking a more friendly error message.   as far as "creator", you can use that also but you would have to use the [semi- private await_() function](https://github.com/sqlalchemy/sqlalchemy/blob/3fd1a52794c5463854fe36cbe97595d8489bbf62/lib/sqlalchemy/util/_concurrency_py3k.py#L81) on the returned awaitable when you invoke the asyncpg.connect() function.   we should document that and probably add an async_creator parameter.

**Reply by zzzeek** [reply]
hmmm we do have a nice error message it just isnt working for that event hook, will fix

**Reply by jrouly** [reply]
Yeah, that asyncio page you linked is indeed what led me to use `engine.sync_engine`, although I have to admit I had encountered that same section probably 3-4 times over the ~5 hours it took me to debug this, and it wasn't until the final reading that it became clear how it related to my problem.

An `async_creator` parameter would be great! And/or documentation around how to use async creator values. I wasn't able to get the `await_` invocation working at all. Although, the event loop troubles I was having _may_ have been related to async pytest rather than SQLAlchemy. Who knows at this point 🙃 

Let me know if you'd like me to file any issues for the new parameter or the friendly error message.
